Invites update

dcavins opened the pull request:
Hello all-

I had to do some updates to make the REST API not throw errors with the new BP_Invitations changes I'm proposing. In doing so, I realized that the endpoints could use some expansion to cover the typical use cases:
Group admin wants to view outstanding invitations to a specific group.
User wants to view/accept/reject his incoming group invitations (to any group).
User wants to view/create/delete his sent group invitations (from any group).

Group admins want to view/approve/reject pending membership requests (for a specific group).
User wants to view/create/delete membership requests (to any group).

Add tests cases to cover each of these scenarios.

The commits do accomplish those changes. I changed the endpoints pretty significantly, though, so I hope the changes will meet with your approval.

One format thing that appears to be normal across REST endpoints, but seemed odd to me is that you can fetch a single invitation at `GET:buddypress/v1/groups/invites/{invite_id}` but when you change the verb to `PUT` or `DELETE` that `{invite_id}` position becomes `{user_id}`. I'd really prefer it stay `{invite_id}` for every verb, and allow deletion or acceptance by `invite_id` or by specifying parameters `group_id`, `user_id`, and optionally `inviter_id`. It would be some work to revisit the logic, but the new classes make finding invitations to work with much easier, and it's way easier to change the incoming parameters before it has been released to the public. :)

Please let me know what you think of my changes and also what you think of my suggestion. I'd be happy to make the changes.

Note that these changes require a BP installation that has my latest changeset from [#6210 ](https://buddypress.trac.wordpress.org/ticket/6210) applied.

Thanks!

imath commented:
Thanks @dcavins I've tried to test your PR but it seems you haven't rebased it regularly to latest version of master during your work. You missed these 26 commits:
- fcf2fbd78d8742190efcf139cf8f4e6cafe897c6
- 72679c928546f29404d5cbf08931bd66d4b232ce
- d95cac4eebc998788ec4779bd3f972bad579006f
- d9d64f04f2517c02831b6b989dc6b281281e6f20
- b75710d4c24a293ff3846ff199c77cdb3fb34c92
- cbdb70b6c3ab74c543b829d4b2276fa46466d574
- 39cd9c37eaf08e0ac46debfe91b3c41290aacf68
- 52a7c230a59b5153d5d07d68fa041b9a40f7e256
- b16ffe1a8834375c273cf961225f268557c4da99
- 203eecb3518b5dfc8d6557941391ee14522274c8
- cc2d9f51f42322c97692e7db864e0d8bc67b358b
- 673252e52adb1ecc84fac71f178410f2d2d2afa7
- 6192e5a294e3a9d5bd8fcea91735cb5b6dbb8b25
- c8858f8497d54a5a0bd8537b0bd2e9f761f1fa6e
- 52534e8d25913694376808cc2c4b883afda0c32e
- f760642b51ad0bdf6f64d3d12e98c5ffa0427fb3
- 68350625d87e45c78c0d7796aede13ecf6d712bb
- be5edf69131f97b4ee24ad46a2169f8a170a4e91
- f8c92731570510eb9fe7adb12157847b2ec708c9
- 2e3f676e8c401bb29122e91fd1757053e5f54c32
- 658b6bc51c9d8929d6d49b9fbacf4af882de8298
- 094880e1c9ff190e519746aa89aa9f2812ecb69e
- 0b95a06e84c8191f284dc805796b84c9b79ce04e
- f574c691bbf0c5691ba3725ed7815c5fcbc78186
- 10b85cfd6fabedc695871288eebd6fd0b130c409
- 7c046dd80946b2079b909198ba95db73ddb031a9

Could you rebase it to latest master and force the push of the PR ?

I've done it locally so I'll be able to test, but if we want to merge I think it would be better.

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -158,38 +179,155 @@ public function get_items_permissions_check( $request ) {
+			 * So, the request must be scoped if the user is not a site admin.
+			 */
+			if ( bp_current_user_can( 'bp_moderate' )
+				|| $group_id_arg && $this->can_see( $group_id_arg ) // @TODO: This allows group admins, too. Is that OK?
```
Yes, it's ok for Group Admins to view the pending invites.

dcavins replied:
👍

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -158,38 +179,155 @@ public function get_items_permissions_check( $request ) {
 			$retval = new WP_Error(
 				'bp_rest_authorization_required',
-				__( 'Sorry, you need to be logged in to see the group invitations.', 'buddypress' ),
+				__( 'Sorry, you need to be logged in to get a membership.', 'buddypress' ),
```
Why changing the message here ?

dcavins replied:
👍

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -158,38 +179,155 @@ public function get_items_permissions_check( $request ) {
 		 * @param WP_REST_Request $request The request sent to the API.
 		 */
-		return apply_filters( 'bp_rest_group_invites_get_items_permissions_check', $retval, $request );
+		return apply_filters( 'bp_rest_group_membership_request_get_item_permissions_check', $retval, $request );
```
Why changing the filter name ? Aren't we into the Group Invites endpoint ?

dcavins replied:
👍

imath commented on tests/membership/test-group-membership-request-controller.php:
```diff
@@ -51,39 +72,84 @@ public function test_get_items() {
+		// $this->create_membership_request( $this->group_id, $u );
+		// $this->create_membership_request( $this->group_id, $u2 );
+		// $this->create_membership_request( $this->group_id, $u3 );
+		groups_send_membership_request( array( 'group_id' => $this->group_id, 'user_id' => $u ) );
```
The 3 above commented lines can probably be removed.

dcavins replied:
👍

imath commented on tests/membership/test-group-membership-request-controller.php:
```diff
@@ -122,20 +192,22 @@ public function test_get_item() {
 		$this->assertFalse( $accepted );
 
-		$this->check_user_data( $user, $data, $member_object );
+		// $this->check_user_data( $user, $data, $member_object );
```
Why is this commented ? Looks like there's no assertions made here ?

dcavins replied:
The important assertion is `$this->assertFalse( $accepted );`

imath commented on tests/membership/test-group-membership-request-controller.php:
```diff
@@ -422,19 +575,20 @@ public function test_prepare_item() {
+	// 	$group_member->save();
+
+	// 	return $group_member;
+	// }
```
Maybe remove completely this method if we don't need it anymore ? Or is there a specific reason to keep it commented ?

dcavins replied:
👍

imath commented:
About your suggestion, when you say:
> It would be some work to revisit the logic, but the new classes make finding invitations to work with much easier, and it's way easier to change the incoming parameters before it has been released to the public

Do you think it can be done before our next dev-chat (August 7th) ?

dcavins commented:
Hi @imath-

Thanks very much for your code review. 🥇 I've rebased from master and added your suggested changes. Git made it look like the messages had been changed, but I actually added the get_item() routine (and its associated permissions check) after working on the membership requests side, so it was copypasta. 

I'm not having any failing tests (in BP or in this REST repo), so I'm guessing the failures you're seeing are due to my not providing up-to-date patches for both sides. I'm updating the BP patch now and will post. Thanks again!

imath commented:
Awesome! Thanks for updating the PR and the patch on Trac-6210. I'll check it asap 👍

dcavins commented:
I've updated the group invites and memberships endpoints to behave more like the WP posts endpoint. Examples:

Fetch invitations where the current user is the invitee:
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/

Fetch invitations to a specific group (must be group admin or site admin):
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?group_id=3

Fetch invitations to a specific user (must be the user or site admin):
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?user_id=3

Fetch invitations from a specific user (must be the user or site admin):
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?inviter_id=3

Find an invitation by a mix of its parameters (must be the inviter, invitee, group admin or site admin):
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?group_id=3&user_id=2
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?group_id=3&inviter_id=6
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/?group_id=3&user_id=2&inviter_id=6

Send an invitation to a user from a group, setting the current user as the inviter:
POST http://bptest.local/wp-json/buddypress/v1/groups/invites/?group_id=3&user_id=2

Fetch an invitation by ID:
GET http://bptest.local/wp-json/buddypress/v1/groups/invites/{invite_id}

Accept an invitation (must be invitee or site admin):
PUT http://bptest.local/wp-json/buddypress/v1/groups/invites/{invite_id}

Delete an invitation (must be invitee, inviter, group admin or site admin ):
DELETE http://bptest.local/wp-json/buddypress/v1/groups/invites/{invite_id}

Fetch requests where the current user is the requestor:
GET http://bptest.local/wp-json/buddypress/v1/groups/membership-requests/

Other requests endpoints: same patterns as the invites endpoint, just no inviter_id parameter.

And it looks like it needs a rebase! Ha ha. Anyway, thanks for looking.

imath commented:
Cool! Thanks a lot @dcavins I'll look at it asap ! Just gave a quick look: it looks promising 💪

dcavins commented:
> Cool! Thanks a lot @dcavins I'll look at it asap ! Just gave a quick look: it looks promising 💪

Reformatting it to a posts-style endpoint was much simpler than the previous approach. However, the new style wasn't possible with the way invites and requests worked before the changes in 6210. So it was a catch 22! :)

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -91,15 +106,20 @@ public function register_routes() {
 		$args = array(
-			'group_id'     => $group->id,
-			'is_confirmed' => $request['is_confirmed'],
+			'item_id'      => isset( $request['group_id'] ) ? $request['group_id'] : false,
```
Those fallbacks should not be necessary as the `get_collection_params` already have the default values as their fallbacks.

dcavins replied:
👍

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
 	 */
 	public function get_items_permissions_check( $request ) {
-		$retval = true;
+		$allow = true;
```
Just for consistency, I'd recommend to change `allow` to `retval`.

dcavins replied:
Ah, if I could go back in time, I'd use meaningful variable names in BuddyPress instead of `$retval`, but it's obviously too late for that. :)

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
+				'bp_rest_authorization_required',
+				__( 'Sorry, you are not allowed to fetch group invitations with those arguments.', 'buddypress' ),
+				array(
+					'status' => rest_authorization_required_code(),
```
This should be a `500` error code.

```suggestion
					'status' => 500,
```

dcavins replied:
👍

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
+		// If the query is not restricted by group or user, limit it to the current user, if not an admin.
+		if ( ! $group_id_arg && ! $user_id_arg && ! bp_current_user_can( 'bp_moderate' ) ) {
+			$user_id_arg = $user_id;
+		}
```
Maybe return early with an error for the group or user.

```
$retval = new WP_Error(
	'bp_rest_invalid_id',
	__( 'Invalid group or member id.', 'buddypress' ),
	array(
		'status' => 404,
	)
);
```

dcavins replied:
This isn't an error, it's a convenience fallback that fills in the current user when in some cases. 👍

renatonascalves replied:
@dcavins I know, what I meant is that we need to check if the user or group exists before checking its permissions.

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
 
-		if ( true === $retval && ! $this->can_see( $group->id ) ) {
-			$retval = new WP_Error(
+		$this->invite = current( $invites );
```
Any reason why this was assigned to this property?

dcavins replied:
I'm not sure--that's how it was done in the code base I started with, so maybe there was some thought of making the invite persist from the permission check to the get() or similar, but the invitations are all cached independently, so it shouldn't be needed.

renatonascalves replied:
I agree! As it is right now, I don't see why it is the way it is.

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
+		} else {
+			$this->invite = new BP_Invitation( 0 );
+		}
+		$allow = true;
```
Return an error here if the id is invalid.

dcavins replied:
👍

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
+				'bp_rest_authorization_required',
+				__( 'Sorry, you are not allowed to fetch an invitation.', 'buddypress' ),
+				array(
+					'status' => rest_authorization_required_code(),
```
```suggestion
					'status' => 500,
```

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -455,10 +705,17 @@ public function delete_item_permissions_check( $request ) {
+			'type'          => $invite->type,
+			'message'       => array(
+				'raw'      => $invite->content,
+				'rendered' => apply_filters( 'the_content', $invite->content ),
```
Are you sure this is the same filter used on BuddyPress core?

dcavins replied:
The invite content is not yet displayed anywhere in a BP template, but I'd guess that `the_content` is probably the right filter. If there is a better filter, please let me know.

renatonascalves replied:
K! I don't know of a better one. If there is no other, I'd agree this is the best one. :)

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -455,10 +705,17 @@ public function delete_item_permissions_check( $request ) {
 	 */
 	public function prepare_item_for_response( $invite, $request ) {
 		$data = array(
-			'user_id'      => $invite->user_id,
```
This method first variable doc is wrong. Should be `BP_Invitation`.

renatonascalves commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -490,7 +747,7 @@ public function prepare_item_for_response( $invite, $request ) {
 	 */
 	protected function prepare_links( $invite ) {
```
PHPDoc wrong: `BP_Invitation` is the correct one.

imath commented:
@dcavins for your info, I've just rebased your branch :) There were twice the same commits so I've cherry picked only one of them.

I've run the phpunit tests using latest patch on BP6210 and I confirm the tests are all ok. I'm going to explore it a little more.

@renatonascalves Thanks a lot for your review 👍

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -148,48 +168,160 @@ public function get_items( $request ) {
-		if ( ! is_user_logged_in() ) {
-			$retval = new WP_Error(
+		if ( ! $user_id ) {
+			$allow = new WP_Error(
```
Another `$allow` to replace by `$retval` 😉

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -271,17 +416,55 @@ public function create_item( $request ) {
+		$group_id_arg   = isset( $request['group_id'] ) ? $request['group_id'] : false;
+		$user_id_arg    = isset( $request['user_id'] ) ? absint( $request['user_id'] ) : false;
+		$inviter_id_arg = isset( $request['inviter_id'] ) ? absint( $request['inviter_id'] ) : bp_loggedin_user_id();
+		$allow         = true;
```
Another `$allow` to replace by a `$retval`

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -293,20 +476,22 @@ public function create_item_permissions_check( $request ) {
 		}
 
-		$accept = groups_accept_invite( $user->ID, $group->id );
+		$this->invite = current( $invites );
```
I agree with @renatonascalves using `$this` shouldn't be necessary or if it is then `$invite` should probably be described at the beginning of the Endpoint Controller

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -351,17 +537,46 @@ public function update_item( $request ) {
+		} else {
+			$this->invite = new BP_Invitation( 0 );
+		}
+		$allow = true;
```
Another `$allow` to replace by `$retval` 😉

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -373,20 +588,24 @@ public function update_item_permissions_check( $request ) {
+		$user         = bp_rest_get_user( $this->invite->user_id );
+		$group        = $this->groups_endpoint->get_group_object( $this->invite->item_id  );
+
+		$deleted = groups_delete_invite( $this->invite->user_id, $this->invite->item_id, $this->invite->inviter_id );
```
Working on 6210, I've noticed `groups_uninvite_user()` was used into template packs Ajax functions. I wonder what's the difference between `groups_delete_invite()` && `groups_uninvite_user()` 🤔

dcavins replied:
The difference is in what hooks are called (and the same goes for `groups_reject_invite()`), but it's confusing to me, too. From the function names, I guess you'd think:
uninvite -> when the inviter or site admin want to delete an invite to a user
reject -> when the invitee wants to delete an invite
delete -> no idea what intent there could be here.

It's a mystery to me, and something I've wonderd about in the past.

imath commented on includes/bp-groups/classes/class-bp-rest-group-invites-endpoint.php:
```diff
@@ -431,17 +650,48 @@ public function delete_item( $request ) {
+		} else {
+			$this->invite = new BP_Invitation( 0 );
+		}
+		$allow = true;
```
Another `$allow` to replace by `$retval`

imath commented on includes/bp-groups/classes/class-bp-rest-group-membership-request-endpoint.php:
```diff
@@ -197,74 +202,73 @@ public function get_items( $request ) {
 	 */
 	public function get_items_permissions_check( $request ) {
-		$retval = true;
+		$allow        = true;
```
Guess what 😄

dcavins replied:
👍

dcavins commented:
I'm closing this PR in favor of a fresh PR with a much simpler commit history. See you there!
